src/websockets.ts

```diff
@@ -0,0 +1,93 @@
+import { websockets_url } from "./Constants";
+import { getPluginId, getPluginName, getVersion, getOs, getOffsetSeconds } from "./Util";
+import { handleAuthenticatedPluginUser } from "./message_handlers/authenticated_plugin_user";
+import { handleIntegrationConnectionSocketEvent } from "./message_handlers/integration_connection";
+import { getItem, getPluginUuid } from './managers/FileManager';
+
+const WebSocket = require("ws");
+
+let retryTimeout = undefined;
+
+export function initializeWebsockets() {
+  const options = {
+    headers: {
+      Authorization: getItem("jwt"),
+      "X-SWDC-Plugin-Id": getPluginId(),
+      "X-SWDC-Plugin-Name": getPluginName(),
+      "X-SWDC-Plugin-Version": getVersion(),
+      "X-SWDC-Plugin-OS": getOs(),
+      "X-SWDC-Plugin-TZ": Intl.DateTimeFormat().resolvedOptions().timeZone,
+      "X-SWDC-Plugin-Offset": getOffsetSeconds() / 60,
+      "X-SWDC-Plugin-UUID": getPluginUuid(),
+    },
+  };
+
+  const ws = new WebSocket(websockets_url, options);
+
+  ws.on("open", function open() {
+    console.debug("[CodeTime] websockets connection open");
+  });
+
+  ws.on("message", function incoming(data) {
+    console.debug("[CodeTime] received websocket message: ", data);
+
+    handleIncomingMessage(data);
+  });
+
+  ws.on("close", function close(code, reason) {
+    console.debug("[CodeTime] websockets connection closed");
+
+    retryConnection();
+  });
+
+  ws.on("unexpected-response", function unexpectedResponse(request, response) {
+    console.debug("[CodeTime] unexpected websockets response:", response.statusCode);
+
+    if (response.statusCode === 426) {
+      console.error("[CodeTime] websockets request had invalid headers. Are you behind a proxy?");
+    } else {
+      retryConnection();
+    }
+  });
+
+  ws.on("error", function error(e) {
+    console.error("[CodeTime] error connecting to websockets", e);
+  });
+}
+
+function retryConnection() {
+  console.debug("[CodeTime] retrying websockets connecting in 10 seconds");
+
+  retryTimeout = setTimeout(() => {
```

Comment:
would this benefit from a max retry count?

Reply:
Yeah, i'm not sure how we should proceed with this. Its the same in code time, but you're right. If we're having issue, it most likely won't ever reconnect. I'll add that. thanks

Reply:
updated, thanks again for reviewing that